validator/client/src/main/java/tech/pegasys/teku/validator/client/DoppelgangerDetection.java

```diff
@@ -0,0 +1,288 @@
+              LOGGER.error(
+                  "Unable to check validators doppelganger. Unable to get validators liveness: {}",
+                  extractErrorMessage(throwable));
+              return null;
+            })
+        .thenApply(doppelgangerDetected -> null);
+  }
+
+  private void checkValidatorDoppelgangers(
+      final Optional<List<ValidatorLivenessAtEpoch>> validatorLivenessAtEpoches,
+      final List<UInt64> validatorIndices,
+      final UInt64 epoch,
+      final UInt64 slot) {
+    List<ValidatorLivenessAtEpoch> liveValidators =
+        filterLiveValidators(validatorLivenessAtEpoches, validatorIndices);
+    if (!liveValidators.isEmpty()) {
+      LOGGER.fatal("Doppelganger detected. Shutting down Validator Client.");
+      this.validatorIndexProvider
+          .getValidatorIndicesByPublicKey()
+          .thenApply(
+              blsPublicKeyIntegerMap -> {
+                STATUS_LOG.validatorsDoppelgangerDetected(
+                    mapLivenessAtEpochToIndicesPubKeysStrings(
+                        liveValidators, blsPublicKeyIntegerMap));
+                stopDoppelgangerDetection(
+                        mapLivenessAtEpochToIndicesPubKeys(
+                            liveValidators, Optional.of(blsPublicKeyIntegerMap)))
+                    .ifExceptionGetsHereRaiseABug();
+                return null;
+              })
+          .exceptionally(
+              error -> {
+                LOGGER.error(
+                    "Unable to get doppelgangers public keys. Only indices are available: {}",
+                    error.getMessage());
+                STATUS_LOG.validatorsDoppelgangerDetected(
+                    liveValidators.stream()
+                        .collect(Collectors.toMap(e -> e.getIndex().intValue(), e -> "")));
+                stopDoppelgangerDetection(
+                        mapLivenessAtEpochToIndicesPubKeys(liveValidators, Optional.empty()))
+                    .ifExceptionGetsHereRaiseABug();
+                return null;
+              })
+          .ifExceptionGetsHereRaiseABug();
+    } else {
+      LOGGER.info("No validators doppelganger detected for epoch {}, slot {}", epoch, slot);
+    }
+  }
+
+  private Map<Integer, String> mapLivenessAtEpochToIndicesPubKeysStrings(
+      List<ValidatorLivenessAtEpoch> liveValidators,
+      Map<BLSPublicKey, Integer> blsPublicKeyIntegerMap) {
+    return blsPublicKeyIntegerMap.entrySet().stream()
+        .filter(
+            pubKeyIndexEntry ->
+                liveValidators.stream()
+                    .anyMatch(
+                        validatorLivenessAtEpoch ->
+                            validatorLivenessAtEpoch
+                                .getIndex()
+                                .equals(UInt64.valueOf(pubKeyIndexEntry.getValue()))))
+        .collect(Collectors.toMap(Map.Entry::getValue, e -> e.getKey().toString()));
+  }
+
+  private Map<Integer, BLSPublicKey> mapLivenessAtEpochToIndicesPubKeys(
```

Comment:
don't you have some code in common with `mapLivenessAtEpochToIndicesPubKeysStrings`?

Reply:
Yeah the filtering step is in common. I extracted it in a separate method...